Bottom: d5e6b50fedd06021794610b39393d86749aa9676
Top:    4532af85c38ffb7afcc8ebf5d0ad3e7f5d31b860
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:44:14 +0800

tuntap: enable bh early during processing XDP

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 2f75e86..3ca0772 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1706,22 +1706,18 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 			goto err_xdp;
 		}
 	}
+	rcu_read_unlock();
+	local_bh_enable();
 
 	skb = build_skb(buf, buflen);
-	if (!skb) {
-		rcu_read_unlock();
-		local_bh_enable();
+	if (!skb)
 		return ERR_PTR(-ENOMEM);
-	}
 
 	skb_reserve(skb, pad - delta);
 	skb_put(skb, len);
 	get_page(alloc_frag->page);
 	alloc_frag->offset += buflen;
 
-	rcu_read_unlock();
-	local_bh_enable();
-
 	return skb;
 
 err_redirect:
